### PR TITLE
chore(docs): patterns for large data sets

### PR DESCRIPTION
**Pull request by tomru**
Adding this here for now until we have a section in the documentation, where we can stick it in.

There has been no feedback on this yet, so happy to hear some here as well.

### CODE REVIEW

File: PATTERNS.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
+# React Patterns
+
+## Using virtual lists/grids for large data sets
+
+When rendering any form of lists in React with lots of elements performance becomes soon an issue. Usually such a list/grid is inside a scrollable container.
```
Comment (larister) [inline PATTERNS.md]:
```suggestion
When rendering any form of lists in React with lots of elements, performance becomes soon an issue. Usually such a list/grid is inside a scrollable container.
```

File: PATTERNS.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+This problem is not React specific, but will bite you earlier in React in my experience.
```
Comment (larister) [inline PATTERNS.md]:
```suggestion
This problem is not React specific, but it'll bite you earlier in React in my experience.
```

File: PATTERNS.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+The common solution outside of the React ecosystem, is to keep the list of things in memory and render just render a subset of the full list of items, i.e. the currently visible items and not render the reset of the list/grid. Whenever the scroll position of the container changes, the items to render need to be updated. This ensures there is just a constant number of DOM elements at any given time and reduces the complexity to draw a list of n elements from `O(n)` to `O(1)`.
```
Comment (larister) [inline PATTERNS.md]:
```suggestion
The common solution outside of the React ecosystem is to keep the list of things in memory and just render a subset of the full list of items, i.e. the currently visible items and not the rest of the list/grid. Whenever the scroll position of the container changes, the items to render need to be updated. This ensures there is just a constant number of DOM elements at any given time and reduces the complexity to draw a list of n elements from `O(n)` to `O(1)`.
```

File: PATTERNS.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+Customers having lots of categories/tags faced this issue in the revamped charts component with the customize bar. There the e.g. Categories are rendered wrappend in a DropdownMenuItem. DropdownMenu creates the scrollable container.
```
Comment (larister) [inline PATTERNS.md]:
```suggestion
Customers with a lot of categories/tags faced this issue in the revamped charts component. In the customize bar on this component, the categories (for example) are wrapped in a `DropdownMenuItem`. The `DropdownMenu` creates the scrollable container.
```

File: PATTERNS.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+After having a look at [react-virtualized](https://github.com/bvaughn/react-virtualized) and implementing the first working version with it, it was raised to me that the author has created [react-window](https://github.com/bvaughn/react-window) as an improved version, stripping away a lot of un-related feature requests and a much smaller build size. It was straight forward to move from the predecessor to the new library.
```
Comment (larister) [inline PATTERNS.md]:
```suggestion
After having a look at [react-virtualized](https://github.com/bvaughn/react-virtualized) and implementing the first working version with it, it was raised to me that the author has created [react-window](https://github.com/bvaughn/react-window) as an improved version, stripping away a lot of unrelated features and having a much smaller build size. It was straight forward to move from the predecessor to the new library.
```

File: PATTERNS.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+In that implementation we've been using [FixedSizeList](https://github.com/bvaughn/react-window/blob/master/src/FixedSizeList.js) and [VariableSizeList](https://github.com/bvaughn/react-window/blob/master/src/VariableSizeList.js). There are corresponding classes for Grids available as well.
+
+### Fixed Size List
+
+FixSizeList is for lists where each element has a fixed size (height or width depending on the direction or the list). Providing `itemSize` and `itemCount` the full size of the container is defined. By providing _renderer_ as a function as children, react-window is able to render any element that is visible. The function is called with `index` and `style`, where `index` is the index of the complete list and `style` stylesheets to position the element absolute on the list in the correct position.
```
Comment (larister) [inline PATTERNS.md]:
```suggestion
`FixedSizeList` is for lists where each element has a fixed size (height or width depending on the direction or the list). The full size of the container is defined by providing `itemSize` and `itemCount`. You can specify a _renderer_ function as children. The renderer function is passed `index` and `style` params, where `index` is the index of the item within the complete list and `style` absolutely positions the element in the list in the correct position.
```

File: PATTERNS.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+   </FixedSizeList>
+ );
+};
+```
+
+[sandbox](https://codesandbox.io/s/xl70pr005p)
+
+## Variable Size List
+
+As the size of an item is not fixed, the `VariableSizeList` must be provided with a function to determine the item size on the fly, whenever a item in the list should be shown (e.g. `index => data[index].size * 5`). So as long as an element has not been shown, the size of the complete list needs to be estimated. There is the prop `estimatedItemSize` to control the size that is reserved for one item in the list. As the size is just estimated the size of the scrollable container may most likely change over time and with the scrollbar might jump a bit when that happens. This is though hardly noticable for long lists.
```
Comment (larister) [inline PATTERNS.md]:
```suggestion
If the size of an item is not fixed, the `VariableSizeList` must be provided with a function to determine the item size on the fly whenever a item in the list should be shown (e.g. `index => data[index].size * 5`). 

As long as an element has not been shown, the size of the complete list needs to be estimated. The prop `estimatedItemSize` is used to control the rough size that is reserved for one item in the list. As the size is just estimated, the size of the scrollable container may change over time, and with the scrollbar might jump a bit when that happens. This is though hardly noticeable for long lists, but to minimize this issue try to make sure the `estimatedItemSize` is as accurate as possible.
```

File: PATTERNS.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+     {Row}
+   </VariableSizeList>
+ );
+};
+```
+
+[sandbox](https://codesandbox.io/s/z2oow6yvyx)
+
+## AutoSizer
+Each of the approaches requires you to set a fixed width and height for the scrolling container.
```
Comment (larister) [inline PATTERNS.md]:
```suggestion
Both of the above approaches require you to set a fixed width and height for the scrolling container.
```

File: PATTERNS.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+`react-virtualized` had a HOC `AutoSizer` that would automatically take all space availabe inside the parent.
```
Comment (larister) [inline PATTERNS.md]:
```suggestion
`react-virtualized` had a Higher Order Component named `AutoSizer` which would automatically take all space available inside the parent.
```

File: PATTERNS.md [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+This component has also been extracted into [react-virtualized-auto-sizer](https://www.npmjs.com/package/react-virtualized-auto-sizer).
```
Comment (larister) [inline PATTERNS.md]:
```suggestion
This component has been extracted into [react-virtualized-auto-sizer](https://www.npmjs.com/package/react-virtualized-auto-sizer) allowing us to use it with react-window.
```

**Comment by tomru** [comment]
> Are there any more docs for react-window? Or blog posts?

There are quite some blog posts out there about the previous package `react-virtualized` of the author. I think most of it probably also applies to the `react-window` though...

**Comment by tomru** [comment]
@larister thanks for these - and sorry for not having sensible line breaks in there. All suggestions addressed and line breaks added now.

**Comment by tomru** [comment]
@larister have added some related blog posts

**Comment by larister** [comment]
Great thanks @tomru. The first blog post mentions docs, but I couldn't find any for react-window other than the examples website... if you know of any hidden docs please could you add them too - otherwise JFMI :)

**Comment by tomru** [comment]
> if you know of any hidden docs please could you add them too - otherwise JFMI :)

Nope, not really. There is rudimentary component property documentation on the example page, e.g. https://react-window.now.sh/#/api/FixedSizeList

**Pull request merged**